examples/stocks/lib/stock_data.dart

```diff
@@ -38,54 +39,59 @@ class Stock {
   }
 }
 
-class StockData {
-  StockData(this._data);
+class StockData extends ChangeNotifier {
+  StockData() {
+    if (actuallyFetchData) {
+      _httpClient = createHttpClient();
+      _fetchNextChunk();
+    }
+  }
+
+  final List<String> _symbols = <String>[];
+  final Map<String, Stock> _stocks = <String, Stock>{};
 
-  final List<List<String>> _data;
+  Iterable<String> get allSymbols => _symbols;
 
-  void appendTo(Map<String, Stock> stocks, List<String> symbols) {
-    for (List<String> fields in _data) {
+  Stock operator [](String symbol) => _stocks[symbol];
+
+  bool get loading => _httpClient != null;
+
+  void add(List<List<String>> data) {
+    for (List<String> fields in data) {
       final Stock stock = new Stock.fromFields(fields);
-      symbols.add(stock.symbol);
-      stocks[stock.symbol] = stock;
+      _symbols.add(stock.symbol);
+      _stocks[stock.symbol] = stock;
     }
-    symbols.sort();
+    _symbols.sort();
+    notifyListeners();
   }
-}
 
-typedef void StockDataCallback(StockData data);
-const int _kChunkCount = 30;
-
-String _urlToFetch(int chunk) {
-  return 'https://domokit.github.io/examples/stocks/data/stock_data_$chunk.json';
-}
+  static const int _kChunkCount = 30;
+  int _nextChunk = 0;
 
-class StockDataFetcher {
-  StockDataFetcher(this.callback) {
-    _httpClient = createHttpClient();
-    _fetchNextChunk();
+  String _urlToFetch(int chunk) {
+    return 'https://domokit.github.io/examples/stocks/data/stock_data_$chunk.json';
   }
 
-  final StockDataCallback callback;
   http.Client _httpClient;
 
   static bool actuallyFetchData = true;
 
-  int _nextChunk = 0;
-
   void _fetchNextChunk() {
-    if (!actuallyFetchData)
-      return;
     _httpClient.get(_urlToFetch(_nextChunk++)).then<Null>((http.Response response) {
       final String json = response.body;
       if (json == null) {
-        print("Failed to load stock data chunk ${_nextChunk - 1}");
-        return null;
+        debugPrint('Failed to load stock data chunk ${_nextChunk - 1}');
+        _httpClient = null;
```

Comment:
Close the http client first? I assume that there's some finalization protocol that closes the connection at GC, but it seems at least tidier to just explicitly close it.